%YAML:1.0

# ROS 接口参数配置：
ros_interface:
  subscribe_message:
    is_use_lidar: true
    is_use_gnss: true
    is_use_sick: true
    is_use_imu: false
    is_use_odom: false
    is_use_encoder: false
    lidar_topic: /points_raw
    lidar_queue_size: 1000000    #10hz 存 27小时数据
    sick_topic: /scan
    sick_queue_size: 1000000    
    gnss_topic: /gps/fix   
    gnss_queue_size: 1000000     
    imu_topic: /imu/data
    imu_queue_size: 1000000      
    odom_topic: /odom
    odom_queue_size: 1000000
    encoder_topic: /encoder_left_right
    encoder_queue_size: 1000000 

  advertise_message:
    is_publish_pcd_map: true
    pcd_map_topic: /lidar_mapping_node/map
    pcd_map_topic_queue_size: 1
    is_publish_pose_registration: true
    pose_registration_topic: /lidar_mapping_node/pose_registration
    pose_registration_queue_size: 1
    is_publish_generate_map_status: true
    generate_map_status_topic: /lidar_mapping_node/generate_map_status
    generate_map_status_queue_size: 5
    occupancy_grid_map:
      is_publish_occupancy_grid_map: true
      occupancy_grid_map_topic: /lidar_mapping_node/grid_map
      occupancy_grid_map_queue_size: 1
      update_distance: 1.0
      resolution: 0.1
      empty_grid_value: 255
      occupancy_grid_value: 0


  advertise_service:
    is_publish_generate_map_service: true
    generate_map_service_name: /generate_map
    is_publish_start_builder_service: true        # 如果设置等待开始建图的 service，则等待开始建图的service后，开始建图，否则，开机自动开始接收数据建图
    start_builder_service_name: /start_builder
    is_publish_save_map_service: true        
    save_map_service_name: /save_map


# 目录参数配置
directory:
  project_directory: /home/alorli/test_data/test/                # 工程文件主目录
  save_map_full_directory: /home/alorli/WORK_SPACE_LCJ/map/      # 定位地图保存路径

  map:
    main_directory: map/
    raw_map: raw_map/
    compensation_map: compensation_map/
    gnss_optimization_map: gnss_optimization_map/
    closeloop_optimization_map: closeloop_optimization_map/
    laserscan_map_directory: laserscan_map/

  velodyne:
    main_directory: velodyne/
    raw_directory: raw/
    compensation_directory: compensation/
    pcd_directory: pcd/
    registration_results_filelist: registration_results_filelist.txt
    g2o_file: g2o_pose.g2o

  sick:
    main_directory: sick/
    raw_directory: raw/
    compensation_directory: compensation/
    pcd_directory: pcd/
    registration_results_filelist: registration_results_filelist.txt
    interpolation_results_filelist: interpolation_results_filelist.txt

  gnss:
    main_directory: gnss/
    gnss_fix_pose_file: gnss_fix_pose.txt
    gnss_init_pose_file: gnss_init_pose.txt

  optimization:
    main_directory: optimization/
    gnss_before_optimization_file: gnss_before_optimization.g2o
    gnss_after_optimization_file: gnss_after_optimization.g2o
    gnss_optimization_results_file: gnss_optimization_results.txt
    closeloop_before_optimization_file: closeloop_before_optimization.g2o
    closeloop_after_optimization_file: closeloop_after_optimization.g2o
    closeloop_optimization_results_file: closeloop_optimization_results.txt

  map_partition:
    output_map_directory: output_map/

  pointcloud_map_arealist:
    arealist_filename: arealist.txt

lidar_mapping_node:
  is_run_gnss_aidded_optimization: false
  is_run_closeloop_optimization: false
  is_generate_raw_map: false
  is_generate_compensation_map: false
  is_generate_gnss_optimization_map: false
  is_generate_closeloop_optimization_map: false
  is_generate_laserscan_map: true
  is_run_map_partition: false
  is_generate_arealist_file: false
  is_delete_project_directory_after_save_map: false   # 保存地图以后，是否删除工程目录，以节省空间，部署到车上时，设置成 true，节省空间



map_builder:
  utm_localization: beijing                      # 选择："beijing" 或者 "changsha" added by lichunjing 2021-08-12
  dual_antenna_parameter:                        # 使用双天线建图时的参数
    is_use_dual_antenna: false
    ranges_main_antenna_to_lidar: 0.66           # 单位：米，主天线(定位天线)到雷达之间的距离
    angle_dual_antenna_to_lidar:  24.6           # 单位:度，双天线基线到主天线到雷达连线之间的角度，逆时针为正，顺时针为负
  msg_parameter:
    pcd_map_voxel_leaf_size: 0.1


  # sick 单线雷达地图生成参数
  map_builder_parameter:
    lidar_n_scans: 16                 # 雷达线数
    vlp_mapping_parameter:
      min_map_horizontal_radius: 3.0  # 3.0
      min_keyframe_distance: 0.5      # 0.5
      num_keyframe_submap: 50         #  50
      min_intensity: 0
      max_intensity: 255
    sick_mapping_parameter:
      min_map_horizontal_radius: 1.0  # 3.0
      min_keyframe_distance: 0.0      # 0.2
      num_keyframe_submap: 500        #  50
      min_intensity: 400              # 300
      max_intensity: 800              # 900
  save_vlp_raw_pcd: true
  save_sick_raw_pcd: true



ndt_registration:
  filter_parameter:
    min_horizontal_radius: 3.0  # 5.0
    min_num_points: 1000        # 1000      参与配准的的最少点的个数, added by lichunjing 2021-04-24
    max_radius: 200.0           # 80.0
    voxel_leaf_size: 1.0        # 2.0
    scale_voxel_leaf_size: 1.1  # 1.1      自适应体素滤波,如果体素滤波尺寸太小,每次将体素滤波尺寸放大 1.1倍 added by lichunjing 2021-04-24


  ndt_parameter:
    transformation_epsilon: 0.01
    step_size: 0.1                 # 默认：0.1
    resolution: 1.0
    max_iterations: 30      

  mapping_parameter:
    min_keyframe_distance: 0.1      #默认：1.0
    map_trim_distance: 50.0
    map_passthrough:
      x_min: -60.0
      x_max: 60.0
      y_min: -60.0
      y_max: 60.0
      z_min: -10.0
      z_max: 30.0


gnss_constraints_builder:
  gnss_constraints_builder_parameter:
    min_delta_gnss_time: 0.07                    # 0.03 for 50hz gnss msg    0.07 for 20hz gnss msg   
    min_num_satellites_add_constraints: 4        # 不同的RTK接收机设置值不同，需要根据每台RTK接收机实测数据设置


gnss_optimization:
  gnss_optimization_parameter:
    lidar_keyframe_distance: 0.0                                # 两个雷达关键帧之间的距离， # 默认:1.0, 设置成 0.0，将使用所有的雷达数据帧进行优化
    use_distance_lidar_pose_estimate_init_angle: 100.0          # 使用多长的轨迹长度去估计初始姿态误差角
    distance_interval_add_prior_gnss_pose: 1.0                  # 使用多长的轨迹长度去估计初始姿态误差角    默认：100
    max_iterations: 100

    # rtk天线在vlp雷达上方的距离，rtk实际高度值减去这个值，是雷达实际所在的高度，也就是gnss优化后，点云地图实际所在的海拔高度
    arm_rtk_antennna_in_vlp: 0.15     # 车载建图设备是 0.15

  lever_arm_parameter:
    use_lever_arm: false
    lever_arm_x: -0.3            # 0.05
    lever_arm_y: 0.0
    lever_arm_z: 0.0


closeloop_constraints_builder:
  closeloop_constraints_builder_parameter:
    num_threads: 6
    min_score: 20                                              # 2.5
    select_interval: 10                                         # 6 在 gnss优化后的位姿中每个几个位姿搜索闭环约束位姿
    min_constraints_time: 10.0                                 # 10.0 闭环搜索时间限制，时间差小于这个值，不进行闭环约束搜索，这个值根据采集地图时的速度设置，
    min_constraints_distance: 0.0                              # 2.0 闭环搜索距离限制，距离差小于这个值，不进行闭环约束搜索
    max_constraints_distance: 20.0                             # 10.0 闭环搜索距离限制，距离差大于这个值，不进行闭环约束搜索
    max_constraints_distance_z: 10.0

closeloop_optimization:
  closeloop_optimization_parameter:
    num_fixed_vertex: 10
    max_iterations: 100
    lidar_odom_translation_weight_xy: 0.3
    lidar_odom_translation_weight_z: 0.3
    lidar_odom_rotation_weight: 0.3


map_partition:
  voxel_filter_size: 0.3
  step_range_x: 20.0
  step_range_y: 20.0



sick_pose_interpolation:
  interpolation_parameter:
    extract_frame_distance: 0.0     #两个用于建图的单线雷达位姿之间的距离
    min_delta_sick_time: 0.2        #两个sick位姿之间的时间差大于这个时间，将不进行位姿插值，这一帧单线雷达数据将不用于建图

  # sick雷达向后装，会扫到车轮，线缆等干扰，外参数为粗略估算
  # transform_sick_in_vlp:
  #   x: 0.1
  #   y: 0.0
  #   z: -0.3
  #   roll: 0.0
  #   pitch: 90.0
  #   yaw: 0.0

  # sick雷达向前装，不会会扫到车轮，线缆等，外参数为粗略测量
  transform_sick_in_vlp:
    x: 0.21
    y: 0.0
    z: -0.2
    roll: 0.0
    pitch: 90.0
    yaw: 0.0
